### PR TITLE
Build FieldMasks for proto Structs and Values

### LINKED ISSUE #1570: FieldMask Generation causes a 400 when a google.protobuf.Struct field has populated values
## 🐛 Bug Report

grpc gateway supports [field mask generation](https://grpc-ecosystem.github.io/grpc-gateway/docs/patch.html).  When the item being used to generate the field mask contains a google.protobuf.Struct with populated values grpc gateway returns a 400.  

## To Reproduce

With the following service definition:

```
service TestService {
	rpc PatchTest (PatchTestRequest) returns (PatchTestResponse) {
		option (google.api.http) = {
      patch: "/test"
      body: "item"
    };
	}
}

message Item {
	string foo = 1;
	google.protobuf.Struct struct = 2;
}

message PatchTestRequest {
	Item item = 1;
	google.protobuf.FieldMask update_mask = 2;
}

message PatchTestResponse {}
```

Run the following test.

```
func TestStructFieldMask(t *testing.T) {
	mux := runtime.NewServeMux()
	pbs.RegisterTestServiceHandlerServer(context.Background(), mux, &pbs.UnimplementedTestServiceServer{})
	req := httptest.NewRequest("PATCH", "http://127.0.0.1/test", strings.NewReader(`{"foo": "val1", "struct": {"key1": "val2"}}`))
	resp := httptest.NewRecorder()
	mux.ServeHTTP(resp, req)
	if resp.Code != http.StatusNotImplemented {
		body, _ := ioutil.ReadAll(resp.Body)
		t.Errorf("Expected Not implemented: Got response %d %q", resp.Code, body)
	}
}
```

## Expected behavior

I expected the test to pass.

## Actual Behavior
The test fails with the following error:

> Expected Not implemented: Got response 400 "{\"code\":3, \"message\":\"could not find field \\\"key1\\\" in \\\"google.protobuf.Struct\\\"\", \"details\":[]}"

## Your Environment

mac os, grpc gateway v2, go version go1.14 darwin/amd64

## Additional Information

For my purpose the ideal behavior would populate the field mask to match the json structure and not depend on the proto structure.  For example a request like this:

```
{
  "key1": "val",
  "key2": {
          "key3":"val2",
          "key4": {"key5": "val3"}
   }
}
```

would result in a field mask containing the following paths: `key1, key2.key3, key2.key4.key5` regardless if the protos looked like
```
message item {
  string key1 =1;
  google.protobuf.Struct key2 = 2;
}
```
or
```
message item {
  string key1 = 1;
  message item2 {
    string key3 = 1;
    message item3 {
      string key5 = 1;
    }
    item3 key4 = 2;
  }
  item2 key2 = 2;
}
```

### PR DESCRIPTION
**Pull request by talanknight**
#### References to other Issues or PRs

Fixes #1570 

#### Have you read the [Contributing Guidelines](https://github.com/grpc-ecosystem/grpc-gateway/blob/master/CONTRIBUTING.md)?

yes

#### Brief description of what is fixed or changed

This PR fixes the ability to build field mask paths from fields that are based on proto Structs and Values which can contain any json object or value respectively.

### CODE REVIEW

File: examples/internal/proto/examplepb/echo_service.proto [hunk 1]
```diff
@@ -69,4 +83,11 @@ service EchoService {
 			delete: "/v1/example/echo_delete"
 		};
 	}
+	// EchoPatch method receives a NonStandardUpdateRequest and returns it.
+	rpc EchoPatch(DynamicMessageUpdate) returns (DynamicMessageUpdate) {
+		option (google.api.http) = {
+			patch: "/v1/example/echo_patch"
+      body: "body"
```
Comment (johanbrandhorst) [inline examples/internal/proto/examplepb/echo_service.proto:90]: I think we ended up mixing tabs and spaces here.

File: runtime/fieldmask.go [hunk 2]
```diff
@@ -92,6 +105,47 @@ func FieldMaskFromRequestBody(r io.Reader, msg proto.Message) (*field_mask.Field
 	return fm, nil
 }
 
+func isDynamicProtoMessage(md protoreflect.MessageDescriptor) bool {
+	return md != nil && (md.FullName() == (&structpb.Struct{}).ProtoReflect().Descriptor().FullName() ||
+		md.FullName() == (&structpb.Value{}).ProtoReflect().Descriptor().FullName())
```
Comment (johanbrandhorst) [inline runtime/fieldmask.go:110]: I appreciate that this is effectively a type safe comparison, but I think it should be ok to just use the string paths for something as stable as this, which removes the import from this file
```suggestion
	return md != nil && (md.FullName() == "google.protobuf.Struct" ||
		md.FullName() == "google.protobuf.Value"
```

See https://github.com/grpc-ecosystem/grpc-gateway/blob/v2/runtime/query.go#L251 for a precedent for this.

**Comment by johanbrandhorst** [comment]
The CI errors imply that some of the files need to be regenerated, see https://github.com/grpc-ecosystem/grpc-gateway/blob/master/CONTRIBUTING.md#i-want-to-regenerate-the-files-after-making-changes

File: runtime/fieldmask.go (hunk 2 above)
Comment (talanknight) [reply runtime/fieldmask.go]: Thanks.  Updated this as requested.

File: examples/internal/proto/examplepb/echo_service.proto (hunk 1 above)
Comment (talanknight) [reply examples/internal/proto/examplepb/echo_service.proto]: Fixed. Thank you.

**Comment by johanbrandhorst** [comment]
Thanks for your contribution!

**Pull request merged**